File: backend/app/models/portal.py
from __future__ import annotations

import enum

from sqlalchemy import Boolean, Enum, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, TimestampMixin, _uuid


class PortalPlatform(str, enum.Enum):
    APPLIED_EPIC = "applied_epic"
    VERTAFORE_AMS360 = "vertafore_ams360"
    VERTAFORE_SIRCON = "vertafore_sircon"
    AMTRUST = "amtrust"
    MARKEL = "markel"
    NATIONWIDE_ES = "nationwide_es"
    CNA = "cna"
    GUIDEWIRE = "guidewire"
    HAWKSOFT = "hawksoft"
    EZLYNX = "ezlynx"
    NOWCERTS = "nowcerts"


class Portal(Base, TimestampMixin):
    __tablename__ = "portals"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    platform: Mapped[PortalPlatform] = mapped_column(
        Enum(PortalPlatform, name="portal_platform", native_enum=False, length=64),
        nullable=False,
        index=True,
    )
    display_name: Mapped[str] = mapped_column(String(255), nullable=False)
    base_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    is_supported: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    risky: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

File: extension/src/types/portal.ts
/**
 * Portal types — mirror of `backend/app/models/portal.py` and
 * `backend/app/schemas/portal.py`.
 *
 * Kept as plain string-literal unions so the values are stable, JSON-safe,
 * and trivially comparable with strings returned by the backend.
 */

export type PortalPlatform =
  | "applied_epic"
  | "vertafore_ams360"
  | "vertafore_sircon"
  | "amtrust"
  | "markel"
  | "nationwide_es"
  | "cna"
  | "guidewire"
  | "hawksoft"
  | "ezlynx"
  | "nowcerts";

export const PORTAL_PLATFORMS: readonly PortalPlatform[] = [
  "applied_epic",
  "vertafore_ams360",
  "vertafore_sircon",
  "amtrust",
  "markel",
  "nationwide_es",
  "cna",
  "guidewire",
  "hawksoft",
  "ezlynx",
  "nowcerts",
] as const;

export function isPortalPlatform(value: unknown): value is PortalPlatform {
  return (
    typeof value === "string" &&
    (PORTAL_PLATFORMS as readonly string[]).includes(value)
  );
}

/**
 * Read projection of a Portal row, mirroring `PortalRead` on the backend.
 */
export interface PortalDescriptor {
  id: string;
  platform: PortalPlatform;
  display_name: string;
  base_url: string | null;
  is_supported: boolean;
  risky: boolean;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface PortalList {
  items: PortalDescriptor[];
  total: number;
}
